src/main/java/com/direwolf20/buildinggadgets/common/building/placement/ConnectedSurface.java

```diff
@@ -0,0 +1,139 @@
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.IBlockAccess;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.function.Function;
+
+public final class ConnectedSurface implements IPlacementSequence {
+
+    /**
+     * @param world           Block access for searching reference
+     * @param searchingCenter Center of the searching region
+     * @param side            Facing to offset from the {@code searchingCenter} to get to the reference region center
+     */
+    public static ConnectedSurface create(IBlockAccess world, BlockPos searchingCenter, EnumFacing side, int range, boolean fuzzy) {
+        Region searchingRegion = Wall.clickedSide(searchingCenter, side, range).getBoundingBox();
+        return create(world, searchingRegion, pos -> pos.offset(side), searchingCenter, side, fuzzy);
+    }
+
+    public static ConnectedSurface create(IBlockAccess world, Region searchingRegion, Function<BlockPos, BlockPos> searching2referenceMapper, BlockPos searchingCenter, EnumFacing side, boolean fuzzy) {
+        return new ConnectedSurface(world, searchingRegion, searching2referenceMapper, searchingCenter, side, fuzzy);
+    }
+
+    private final IBlockAccess world;
+    private final Region searchingRegion;
+    private final Function<BlockPos, BlockPos> searching2referenceMapper;
+    private final BlockPos searchingCenter;
+    private final EnumFacing side;
+    private final boolean fuzzy;
+
+    @VisibleForTesting
+    private ConnectedSurface(IBlockAccess world, Region searchingRegion, Function<BlockPos, BlockPos> searching2referenceMapper, BlockPos searchingCenter, EnumFacing side, boolean fuzzy) {
+        this.world = world;
+        this.searchingRegion = searchingRegion;
+        this.searching2referenceMapper = searching2referenceMapper;
+        this.searchingCenter = searchingCenter;
+        this.side = side;
+        this.fuzzy = fuzzy;
+    }
+
+    /**
+     * The bounding box of the searchingRegion that is being searched.
+     */
+    @Override
+    public Region getBoundingBox() {
+        return searchingRegion;
+    }
+
+    /**
+     * {@inheritDoc}<br>
+     * <b>inaccurate representation (case 2)</b>:
+     *
+     * @return {@code getBoundingBox().contains(x, y, z)} since it would be costly to check each position
+     */
+    @Override
+    public boolean mayContain(int x, int y, int z) {
+        return searchingRegion.mayContain(x, y, z);
+    }
+
+    /**
+     * @deprecated ConnectedSurface should be immutable, so this is not needed
+     */
+    @Deprecated
+    @Override
+    public IPlacementSequence copy() {
+        return new ConnectedSurface(world, searchingRegion, searching2referenceMapper, searchingCenter, side, fuzzy);
+    }
+
+    /**
+     * @implNote technically BFS fits better with a flood fill algorithm and it simpler to implement with a generator
```

Comment:
This does not belong as a JavaDoc comment. Please describe what kind of behaviour can be expected from this iterator.